### PR TITLE
ci: images tool 

### PR DESCRIPTION
**Pull request by faiq**
**What problem does this PR solve?**:
Adds a tool to fetch all images CAREN uses

**Which issue(s) this PR fixes**:
Fixes #

**How Has This Been Tested?**:

I ran the following command from `hack/tools/fetch-images`

```
$ go run main.go -chart-directory=../../../charts/cluster-api-runtime-extensions-nutanix/ --helm-chart-configmap=../../../charts/cluster-api-runtime-extensions-nutanix/templates/helm-config.yaml -caren-version=v0.13.1
ghcr.io/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix:v0.13.1
ghcr.io/nutanix-cloud-native/caren-helm-reg:v0.13.1
registry.k8s.io/sig-storage/snapshot-controller:v8.0.1
registry.k8s.io/provider-aws/cloud-controller-manager:v1.27.7
registry.k8s.io/provider-aws/cloud-controller-manager:v1.28.6
registry.k8s.io/provider-aws/cloud-controller-manager:v1.29.3
registry.k8s.io/provider-aws/cloud-controller-manager:v1.30.1
public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.33.0
public.ecr.aws/eks-distro/kubernetes-csi/node-driver-registrar:v2.11.0-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/livenessprobe:v2.13.0-eks-1-30-10
public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.33.0
public.ecr.aws/eks-distro/kubernetes-csi/external-provisioner:v5.0.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/external-attacher:v4.6.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/external-resizer:v1.11.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/livenessprobe:v2.13.0-eks-1-30-10
registry.k8s.io/autoscaling/cluster-autoscaler:v1.30.0
quay.io/metallb/speaker:v0.14.8
quay.io/frrouting/frr:9.1.0
quay.io/metallb/speaker:v0.14.8
quay.io/metallb/controller:v0.14.8
registry.k8s.io/nfd/node-feature-discovery:v0.16.1-minimal
registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.9.3
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-provisioner:v3.6.3
registry.k8s.io/sig-storage/csi-attacher:v4.4.3
registry.k8s.io/sig-storage/csi-resizer:v1.9.3
registry.k8s.io/sig-storage/csi-snapshotter:v3.0.3
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-external-health-monitor-controller:v0.10.0
docker.io/nutanix/ntnx-csi-precheck:3.0.0
quay.io/cilium/cilium:v1.16.0
quay.io/cilium/cilium-envoy@sha256:bd5ff8c66716080028f414ec1cb4f7dc66f40d2fb5a009fff187f4a9b90b566b
quay.io/cilium/operator-generic:v1.16.0
quay.io/cilium/certgen:v0.2.0
docker.io/rancher/local-path-provisioner:v0.0.28
ghcr.io/nutanix-cloud-native/cloud-provider-nutanix/controller:v0.3.2
quay.io/tigera/operator:v1.34.0
```

**Special notes for your reviewer**:

Depends on #896.

### CODE REVIEW

File: hack/tools/fetch-images/main.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		images.RepoURL = info.repo
+	}
+	if info.valuesFile != "" {
+		images.ValueFiles.Set(info.valuesFile)
+	}
+	if len(info.stringValues) > 0 {
+		images.StringValues = info.stringValues
+	}
+	// kubeVersion needs to be set for some addons
+	images.KubeVersion = "v1.29.0"
```
Comment (dlipovetsky) [inline hack/tools/fetch-images/main.go:227]: Does this need to change over time? If so, should the command caller pass it as a flag?

Comment (faiq) [reply hack/tools/fetch-images/main.go:227]: there was a minimum version required for an app.

File: .goreleaser.yml [hunk 2]
```diff
@@ -1,9 +1,9 @@
 # Copyright 2023 Nutanix. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
-version: 2
+# yaml-language-server: $schema=https://goreleaser.com/static/schema.json
 
-project_name: cluster-api-runtime-extensions-nutanix
```
Comment (supershal) [inline .goreleaser.yml:6]: did this removed accidentally? goreleaser can inferred it. however this change seems unrelated.

File: hack/tools/fetch-images/main.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"slices"
+	"strings"
+	"text/template"
+
+	"github.com/d2iq-labs/helm-list-images/pkg"
```
Comment (supershal) [inline hack/tools/fetch-images/main.go:17]: nit: it would be good to use alias here.

File: hack/tools/fetch-images/main.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+		os.Exit(1)
+	}
+	if chartDirectory == "" || helmChartConfigMap == "" {
+		fmt.Println("chart-directory helm-chart-configmap must be set")
+		os.Exit(1)
+	}
+	i := &ChartInfo{
+		name: chartDirectory,
+	}
+	if carenVersion != "" {
```
Comment (supershal) [inline hack/tools/fetch-images/main.go:73]: should make carenVersion required flag.

File: hack/flakes/flake.nix [hunk 5]
```diff
@@ -162,9 +162,44 @@
             '';
           };
 
+          helm-list-images = buildGo122Module rec {
```
Comment (supershal) [inline hack/flakes/flake.nix:165]: Do we need this plugin anymore? Wont it be downloaded as go dependency of the tool?

**Review by supershal — COMMENTED** [review]
Run in locally.
```
❯ cat caren-images.txt
ghcr.io/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix:v0.0.0-dev
ghcr.io/nutanix-cloud-native/caren-helm-reg:v0.0.0-dev
quay.io/metallb/speaker:v0.14.8
quay.io/frrouting/frr:9.1.0
quay.io/metallb/speaker:v0.14.8
quay.io/metallb/controller:v0.14.8
registry.k8s.io/nfd/node-feature-discovery:v0.16.1-minimal
ghcr.io/nutanix-cloud-native/cloud-provider-nutanix/controller:v0.4.0
registry.k8s.io/sig-storage/snapshot-controller:v8.0.1
quay.io/tigera/operator:v1.34.3
registry.k8s.io/provider-aws/cloud-controller-manager:v1.28.6
registry.k8s.io/provider-aws/cloud-controller-manager:v1.29.3
registry.k8s.io/provider-aws/cloud-controller-manager:v1.30.1
registry.k8s.io/provider-aws/cloud-controller-manager:v1.27.7
public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.33.0
public.ecr.aws/eks-distro/kubernetes-csi/node-driver-registrar:v2.11.0-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/livenessprobe:v2.13.0-eks-1-30-10
public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.33.0
public.ecr.aws/eks-distro/kubernetes-csi/external-provisioner:v5.0.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/external-attacher:v4.6.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/external-resizer:v1.11.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/livenessprobe:v2.13.0-eks-1-30-10
docker.io/rancher/local-path-provisioner:v0.0.28
quay.io/cilium/cilium:v1.16.1
quay.io/cilium/cilium-envoy@sha256:bd5ff8c66716080028f414ec1cb4f7dc66f40d2fb5a009fff187f4a9b90b566b
quay.io/cilium/operator-generic:v1.16.1
quay.io/cilium/certgen:v0.2.0
registry.k8s.io/autoscaling/cluster-autoscaler:v1.30.0
registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.9.3
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-provisioner:v3.6.3
registry.k8s.io/sig-storage/csi-attacher:v4.4.3
registry.k8s.io/sig-storage/csi-resizer:v1.9.3
registry.k8s.io/sig-storage/csi-snapshotter:v3.0.3
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-external-health-monitor-controller:v0.10.0
docker.io/nutanix/ntnx-csi-precheck:3.0.0
```
Have you already tested this with konvoy e2e airgap tests?

File: hack/tools/fetch-images/main.go (hunk 4 above)
Comment (faiq) [reply hack/tools/fetch-images/main.go:75]: it gets run by the make target so i think it will be ok

**Comment by faiq** [comment]
works 

```
11:58 AM  fraza @ Linux-PC1RG42G  ~/go/src/github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix  faiq/get-images? 
$ make list-images

11:59 AM  fraza @ Linux-PC1RG42G  ~/go/src/github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix  faiq/get-images? 
$ cat caren-images.txt 
ghcr.io/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix:v0.0.0-dev
ghcr.io/nutanix-cloud-native/caren-helm-reg:v0.0.0-dev
public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.34.0
public.ecr.aws/eks-distro/kubernetes-csi/node-driver-registrar:v2.11.0-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/livenessprobe:v2.13.0-eks-1-30-10
public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver:v1.34.0
public.ecr.aws/eks-distro/kubernetes-csi/external-provisioner:v5.0.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/external-attacher:v4.6.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/external-resizer:v1.11.1-eks-1-30-10
public.ecr.aws/eks-distro/kubernetes-csi/livenessprobe:v2.13.0-eks-1-30-10
quay.io/cilium/cilium:v1.16.1
quay.io/cilium/cilium-envoy@sha256:bd5ff8c66716080028f414ec1cb4f7dc66f40d2fb5a009fff187f4a9b90b566b
quay.io/cilium/operator-generic:v1.16.1
quay.io/cilium/certgen:v0.2.0
registry.k8s.io/nfd/node-feature-discovery:v0.16.1-minimal
registry.k8s.io/sig-storage/snapshot-controller:v8.0.1
quay.io/tigera/operator:v1.34.3
registry.k8s.io/provider-aws/cloud-controller-manager:v1.30.2
registry.k8s.io/provider-aws/cloud-controller-manager:v1.31.0
registry.k8s.io/provider-aws/cloud-controller-manager:v1.27.9
registry.k8s.io/provider-aws/cloud-controller-manager:v1.28.9
registry.k8s.io/provider-aws/cloud-controller-manager:v1.29.6
registry.k8s.io/autoscaling/cluster-autoscaler:v1.30.0
docker.io/rancher/local-path-provisioner:v0.0.28
quay.io/metallb/speaker:v0.14.8
quay.io/frrouting/frr:9.1.0
quay.io/metallb/speaker:v0.14.8
quay.io/metallb/controller:v0.14.8
ghcr.io/nutanix-cloud-native/cloud-provider-nutanix/controller:v0.4.1
registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.9.3
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-provisioner:v3.6.3
registry.k8s.io/sig-storage/csi-attacher:v4.4.3
registry.k8s.io/sig-storage/csi-resizer:v1.9.3
registry.k8s.io/sig-storage/csi-snapshotter:v3.0.3
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-external-health-monitor-controller:v0.10.0
docker.io/nutanix/ntnx-csi-precheck:3.0.0
```

File: make/addons.mk [hunk 6]
```diff
@@ -84,3 +84,11 @@ generate-mindthegap-repofile: generate-helm-configmap ; $(info $(M) generating h
...
 template-helm-repository: generate-mindthegap-repofile ## this is used by gorealeaser to set the helm value to this.
 	sed -i '/RepositoryURL:/s#\(RepositoryURL: *\)\(.*\)#\1{{ if .Values.selfHostedRegistry }}oci://helm-repository.{{ .Release.Namespace }}.svc/charts{{ else }}\2{{ end }}#'  "./charts/cluster-api-runtime-extensions-nutanix/templates/helm-config.yaml"
+
+.PHONY: list-images
+list-images:
+	cd hack/tools/fetch-images && go build
+	./hack/tools/fetch-images/fetch-images \
+		-chart-directory=./charts/cluster-api-runtime-extensions-nutanix/ \
+		-helm-chart-configmap=./charts/cluster-api-runtime-extensions-nutanix/templates/helm-config.yaml \
+		-caren-version=$(CAREN_VERSION) >> caren-images.txt
```
Comment (dkoshkin) [inline make/addons.mk:94]: What do you think about letting the caller decide what to do with the output? 
```suggestion
		-caren-version=$(CAREN_VERSION)
```

File: make/addons.mk [hunk 7]
```diff
@@ -84,3 +84,11 @@ generate-mindthegap-repofile: generate-helm-configmap ; $(info $(M) generating h
 .PHONY: template-helm-repository
 template-helm-repository: generate-mindthegap-repofile ## this is used by gorealeaser to set the helm value to this.
 	sed -i '/RepositoryURL:/s#\(RepositoryURL: *\)\(.*\)#\1{{ if .Values.selfHostedRegistry }}oci://helm-repository.{{ .Release.Namespace }}.svc/charts{{ else }}\2{{ end }}#'  "./charts/cluster-api-runtime-extensions-nutanix/templates/helm-config.yaml"
+
+.PHONY: list-images
+list-images:
+	cd hack/tools/fetch-images && go build
+	./hack/tools/fetch-images/fetch-images \
```
Comment (dkoshkin) [inline make/addons.mk:91]: Dropping the redirect would allow for this:
```suggestion
	cd hack/tools/fetch-images && run . \
```

File: hack/tools/fetch-images/main.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+				images = append(images, chartImages...)
+			}
+			// skip the to next addon because we got what we needed
+			continue
+		}
+		chartImages, err := getImagesForChart(info)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get images for %s with error %w", info.name, err)
+		}
+		images = append(images, chartImages...)
```
Comment (dkoshkin) [inline hack/tools/fetch-images/main.go:177]: I see there are some duplicate images being generated, what should be responsible for filtering those?
```
...
docker.io/nutanix/ntnx-csi:3.0.0
registry.k8s.io/sig-storage/livenessprobe:v2.11.0
registry.k8s.io/sig-storage/csi-provisioner:v3.6.3
registry.k8s.io/sig-storage/csi-attacher:v4.4.3
registry.k8s.io/sig-storage/csi-resizer:v1.9.3
registry.k8s.io/sig-storage/csi-snapshotter:v3.0.3
docker.io/nutanix/ntnx-csi:3.0.0
...
quay.io/metallb/speaker:v0.14.8
quay.io/frrouting/frr:9.1.0
quay.io/metallb/speaker:v0.14.8
...

File: hack/tools/fetch-images/main.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+	return fullPath, nil
+}
+
+type HelmChartFromConfigMap struct {
+	Version    string `yaml:"ChartVersion"`
+	Repository string `yaml:"RepositoryURL"`
+	ChartName  string `yaml:"ChartName"`
+}
+
+func getImagesForAddons(helmChartConfigMap, carenChartDirectory string) ([]string, error) {
```
Comment (dkoshkin) [inline hack/tools/fetch-images/main.go:121]: Currently Calico uses an operator that then pulls additional images.
There is some additional tooling that generates a list of images.

Do you think its the responsibility of this tool to have special handling for certain addons? (just Calico for now)
```
calico_cni_images_for_version() {
  local -r calico_operator_version="${1}"
  local -r calico_versions="$(curl -s https://raw.githubusercontent.com/tigera/operator/"${calico_operator_version}"/config/calico_versions.yml)"
  local -r calico_registry="${2:-docker.io}"
  # TODO Parse the images and versions from https://github.com/tigera/operator/blob/${calico_operator_version}/pkg/components/calico.go; it appears to be a better source, because it has the exact image names.
  local -a calico_images=("quay.io/tigera/operator:${calico_operator_version}")
  calico_images+=("${calico_registry}/calico/cni:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/cni"].version')")
  calico_images+=("${calico_registry}/calico/kube-controllers:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/kube-controllers"].version')")
  calico_images+=("${calico_registry}/calico/node:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/node"].version')")
  calico_images+=("${calico_registry}/calico/apiserver:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/apiserver"].version')")
  calico_images+=("${calico_registry}/calico/pod2daemon-flexvol:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output .components.flexvol.version)")
  calico_images+=("${calico_registry}/calico/typha:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output .components.typha.version)")
  calico_images+=("${calico_registry}/calico/csi:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/csi"].version')")
  calico_images+=("${calico_registry}/calico/node-driver-registrar:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["csi-node-driver-registrar"].version')")
  # The `calico/ctl` image contains the Calico CLI, which can be useful for debugging.
  calico_images+=("${calico_registry}/calico/ctl:$(echo "${calico_versions}" | ${GOJQ_BIN} --yaml-input --raw-output '.components["calico/cni"].version')")

  IMAGES=("${IMAGES[@]}" "${calico_images[@]}")
}
```

File: .gitignore [hunk 10]
```diff
@@ -47,3 +47,4 @@ node_modules/
 /_artifacts/
 test/e2e/config/caren-envsubst.yaml
 /release-metadata.yaml
+hack/tools/fetch-images/fetch-images
```
Comment (dkoshkin) [inline .gitignore:50]: Lets add `caren-images.txt` here too.
I'm pretty sure Goreleaser will complain about a dirty git state.

File: charts/cluster-api-runtime-extensions-nutanix/addons/nfd/values-template.yaml [hunk 11]
```diff
@@ -1,3 +1,5 @@
+image:
+  tag: v0.16.1-minimal
```
Comment (jimmidyson) [inline charts/cluster-api-runtime-extensions-nutanix/addons/nfd/values-template.yaml:2]: This is not necessary since nfd 0.15 where it uses a minimal scratch based image by default (also this downgraded the image btw).

File: hack/tools/fetch-images/main.go (hunk 8 above)
Comment (faiq) [reply hack/tools/fetch-images/main.go:178]: updated and it should remove duplicates

File: make/addons.mk (hunk 6 above)
Comment (faiq) [reply make/addons.mk:94]: its only needed by the release process so this isn't necessary.

File: hack/tools/fetch-images/main.go (hunk 9 above)
Comment (jimmidyson) [reply hack/tools/fetch-images/main.go:122]: The helm list images fork that we use here has the capability to add extra images that cannot be discovered automatically

See https://github.com/d2iq-labs/helm-list-images/blob/f30cfc9a97ff64f08badcf12082d1a7d3b78c966/docs/doc/list-images.md?plain=1#L27 for details with an example extra file at https://github.com/d2iq-labs/helm-list-images/blob/f30cfc9a97ff64f08badcf12082d1a7d3b78c966/extra.yaml#L3 (set how you can use helm values that will be templated here).

Comment (faiq) [reply hack/tools/fetch-images/main.go:122]: These images already get accounted for by a process we have to collect all of the cluster resource set images. I don't think it's strictly necessary to handle this right now.

Comment (dkoshkin) [reply hack/tools/fetch-images/main.go:122]: > These images already get accounted for by a process we have to collect all of the cluster resource set images.

Is that some process in this repo?

Comment (faiq) [reply hack/tools/fetch-images/main.go:122]: No, it's in a downstream dependency of this repo.

Comment (jimmidyson) [reply hack/tools/fetch-images/main.go:122]: It seems preferable that this project lists all the images required for all addons deployed by this project. We can do it as a follow up though, I don't mind adding the extra images in a follow up PR.

Comment (faiq) [reply hack/tools/fetch-images/main.go:122]: sounds good- I couldn’t tell what the original ask was when this comment was written. “responsibility” could be argued either way.

Comment (jimmidyson) [reply hack/tools/fetch-images/main.go:122]: Pushed a PR in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/900.

**Review by jimmidyson — DISMISSED** [review]
Great work!

**Review by dkoshkin — DISMISSED** [review]

**Review by supershal — DISMISSED** [review]

**Review by jimmidyson — APPROVED** [review]
Nice!

**Review by mhrabovcin — APPROVED** [review]
